c++/src/sargs/ExpressionTree.cc

```diff
@@ -0,0 +1,192 @@
+  ExpressionTree::ExpressionTree(Operator op)
+                                : mOperator(op)
+                                , mLeaf(UNUSED_LEAF)
+                                , mConstant(TruthValue::YES_NO_NULL) {
+  }
+
+
+  ExpressionTree::ExpressionTree(Operator op,
+                                 std::initializer_list<TreeNode> children)
+                                : mOperator(op)
+                                , mChildren(children.begin(), children.end())
+                                , mLeaf(UNUSED_LEAF)
+                                , mConstant(TruthValue::YES_NO_NULL) {
+    // PASS
+  }
+
+  ExpressionTree::ExpressionTree(size_t leaf)
+                                : mOperator(Operator::LEAF)
+                                , mChildren()
+                                , mLeaf(leaf)
+                                , mConstant(TruthValue::YES_NO_NULL) {
+    // PASS
+  }
+
+  ExpressionTree::ExpressionTree(TruthValue constant)
+                                : mOperator(Operator::CONSTANT)
+                                , mChildren()
+                                , mLeaf(UNUSED_LEAF)
+                                , mConstant(constant) {
+    // PASS
+  }
+
+  ExpressionTree::ExpressionTree(const ExpressionTree& other)
+                                : mOperator(other.mOperator)
+                                , mLeaf(other.mLeaf)
+                                , mConstant(other.mConstant) {
+    for (TreeNode child : other.mChildren) {
+      mChildren.emplace_back(std::make_shared<ExpressionTree>(*child));
+    }
+  }
+
+  ExpressionTree::Operator ExpressionTree::getOperator() const {
+    return mOperator;
+  }
+
+  const std::vector<TreeNode>& ExpressionTree::getChildren() const {
+    return mChildren;
+  }
+
+  std::vector<TreeNode>& ExpressionTree::getChildren() {
+    return const_cast<std::vector<TreeNode>&>(
+      const_cast<const ExpressionTree *>(this)->getChildren());
+  }
+
+  const TreeNode ExpressionTree::getChild(size_t i) const {
+    return mChildren.at(i);
+  }
+
+  TreeNode ExpressionTree::getChild(size_t i) {
+    return std::const_pointer_cast<ExpressionTree>(
+      const_cast<const ExpressionTree *>(this)->getChild(i));
+  }
+
+  TruthValue ExpressionTree::getConstant() const {
+    assert(mOperator == Operator::CONSTANT);
+    return mConstant;
+  }
+
+  size_t ExpressionTree::getLeaf() const {
+    assert(mOperator == Operator::LEAF);
+    return mLeaf;
+  }
+
+  void ExpressionTree::setLeaf(size_t leaf) {
+    assert(mOperator == Operator::LEAF);
+    mLeaf = leaf;
+  }
+
+  void ExpressionTree::addChild(TreeNode child) {
+    mChildren.push_back(child);
+  }
+
+  TruthValue ExpressionTree::evaluate(
+                                 const std::vector<TruthValue>& leaves) const {
+    TruthValue result;
+    switch (mOperator) {
+      case Operator::OR:
+      {
+        result = mChildren.at(0)->evaluate(leaves);
+        for (size_t i = 1; i < mChildren.size(); ++i) {
```

Comment:
Is it better to introduce for short-circuit evaluation for `AND` and `OR` operators?

Reply:
Good catch, will do.